### PR TITLE
WFLY-13520 Add support for Infinispan query, counters, locks, and CDI

### PR DESCRIPTION
**Pull request by rhusar**
Resolves
https://issues.redhat.com/browse/WFLY-13520

Proposal 
https://github.com/wildfly/wildfly-proposals/pull/467

### CODE REVIEW

**Review by pferraro — CHANGES_REQUESTED** [review]
Overall, this looks great.  My comments are largely very minor.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/counter/deployment/InfinispanCounterServlet.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    private static final String SERVLET_NAME = "counter";
+    static final String SERVLET_PATH = "/" + SERVLET_NAME;
+    public static final String COUNTER_TYPE_PARAMETER = "counter-type";
+    public static final String COUNTER_NAME_PARAMETER = "counter-name";
+    public static final String STORAGE_PARAMETER = "define";
+
+    @Resource(lookup = "java:jboss/infinispan/container/counter")
+    private EmbeddedCacheManager cm;
+
+    private CounterManager ecm;
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/counter/deployment/InfinispanCounterServlet.java]: Haha, fixed.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/counter/deployment/web.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+<web-app xmlns="http://java.sun.com/xml/ns/javaee"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd"
+         version="3.0">
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/counter/deployment/web.xml]: Removed.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/counter/InfinispanCounterTestCase.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            try {
+                Assert.assertEquals(HttpServletResponse.SC_OK, response.getStatusLine().getStatusCode());
+                Assert.assertEquals(5, Integer.parseInt(EntityUtils.toString(response.getEntity())));
+            } finally {
+                HttpClientUtils.closeQuietly(response);
+            }
+        }
+    }
+
+    public static class ServerSetupTask extends CLIServerSetupTask {
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/counter/InfinispanCounterTestCase.java]: Done.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/counter/InfinispanCounterTestCase.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        this.test(Storage.PERSISTENT, baseURL1, baseURL2);
+    }
+
+    public void test(Storage storage, URL baseURL1, URL baseURL2) throws IOException, URISyntaxException {
+
+        String counterName = storage.name() + UUID.randomUUID();
+
+        try (CloseableHttpClient client = TestHttpClientUtils.promiscuousCookieHttpClient()) {
+            HttpResponse response = client.execute(new HttpGet(InfinispanCounterServlet.createURI(baseURL1, counterName, storage.name())));
+            try {
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/counter/InfinispanCounterTestCase.java]: Done.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/lock/InfinispanLockTestCase.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    @Test
+    public void test(@ArquillianResource(InfinispanLockServlet.class) @OperateOnDeployment(DEPLOYMENT_1) URL baseURL1,
+                     @ArquillianResource(InfinispanLockServlet.class) @OperateOnDeployment(DEPLOYMENT_2) URL baseURL2)
+            throws IOException, URISyntaxException {
+
+        String lockName = UUID.randomUUID().toString();
+
+        try (CloseableHttpClient client = TestHttpClientUtils.promiscuousCookieHttpClient()) {
+            HttpResponse response = client.execute(new HttpGet(InfinispanLockServlet.createURI(baseURL1, lockName, LockOperation.DEFINE)));
+            try {
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/lock/InfinispanLockTestCase.java]: Done.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/lock/InfinispanLockTestCase.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                Assert.assertEquals(HttpServletResponse.SC_OK, response.getStatusLine().getStatusCode());
+                Assert.assertFalse(Boolean.parseBoolean(EntityUtils.toString(response.getEntity())));
+            } finally {
+                HttpClientUtils.closeQuietly(response);
+            }
+        }
+    }
+
+    public static class ServerSetupTask extends CLIServerSetupTask {
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/lock/InfinispanLockTestCase.java]: Done.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/single/infinispan/embedded/EmbeddedCacheGenericMarshallerTestCase.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import org.jboss.shrinkwrap.descriptor.api.spec.se.manifest.ManifestDescriptor;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * Adopted and adapted from Infinispan testsuite.
+ *
+ *
+ * wont work - relies on org.infinispan.jboss-marshalling; remopved from deps
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/single/infinispan/embedded/EmbeddedCacheGenericMarshallerTestCase.java]: Yes, removed.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/single/infinispan/query/ContainerRemoteQueryTestCase.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+ * Test for remote query using Container-managed objects.
+ *
+ * @author Radoslav Husar
+ * @since 27
+ */
+@RunWith(Arquillian.class)
+@ServerSetup({ContainerRemoteQueryTestCase.ServerSetupTask.class})
+public class ContainerRemoteQueryTestCase {
+
+    public static class ServerSetupTask extends CLIServerSetupTask {
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/single/infinispan/query/ContainerRemoteQueryTestCase.java]: Done.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/single/infinispan/query/ContainerManagedHotRodClientTestCase.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                .addClass(ContainerManagedHotRodClientTestCase.class)
+                .addClass(Person.class)
+                .addClass(PersonSerializationContextInitializer.class)
+                .addClass(PersonSerializationContextInitializer.class.getName() + "Impl")
+                .addAsServiceProvider(SerializationContextInitializer.class.getName(), PersonSerializationContextInitializer.class.getName() + "Impl")
+                .add(new StringAsset(Descriptors.create(ManifestDescriptor.class).attribute("Dependencies", "org.infinispan, org.infinispan.commons, org.infinispan.client.hotrod, org.infinispan.query, org.infinispan.protostream").exportAsString()), "META-INF/MANIFEST.MF")
+                ;
+    }
+
+    public static class ServerSetupTask extends CLIServerSetupTask {
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/single/infinispan/query/ContainerManagedHotRodClientTestCase.java]: Done.

File: testsuite/integration/clustering/src/test/resources/beans.xml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd"
+       version="1.2" bean-discovery-mode="all">
+</beans>
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/resources/beans.xml]: Done.

File: testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/lock/deployment/web.xml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+<web-app xmlns="http://java.sun.com/xml/ns/javaee"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/web-app_3_0.xsd"
+         version="3.0">
+</web-app>
```
[the remark this answers is not in the document]
Comment (rhusar) [reply testsuite/integration/clustering/src/test/java/org/jboss/as/test/clustering/cluster/infinispan/lock/deployment/web.xml]: Removed.

File: ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/infinispan/query/main/module.xml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+    <dependencies>
+        <module name="javax.transaction.api"/>
+
+        <module name="org.apache.lucene"/>
+        <module name="org.hibernate.commons-annotations"/>
+        <module name="org.hibernate.search.engine"/>
+        <module name="org.infinispan"/>
+        <module name="org.infinispan.commons"/>
+        <module name="org.infinispan.lock"/>
```
[the remark this answers is not in the document]
Comment (rhusar) [reply ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/infinispan/query/main/module.xml:37]: Yes, this is used by MassIndexer (Component that rebuilds the Lucene index from the cached data).

File: ee-feature-pack/galleon-common/src/main/resources/layers/standalone/web-clustering/layer-spec.xml [hunk 13]
```diff
@@ -15,5 +15,13 @@
...
              combination requires the clustering<->undertow integration package -->
         <package name="org.wildfly.clustering.web.undertow"/>
+        <!-- This layer also brings in Infinispan modules -->
+        <package name="org.infinispan.cdi.common"/>
+        <package name="org.infinispan.cdi.embedded"/>
+        <package name="org.infinispan.cdi.remote"/>
+        <package name="org.infinispan.counter"/>
+        <package name="org.infinispan.lock"/>
+        <package name="org.infinispan.query"/>
+        <package name="org.infinispan.query.core"/>
```
Comment (pferraro) [inline ee-feature-pack/galleon-common/src/main/resources/layers/standalone/web-clustering/layer-spec.xml:25]: Is it possible to instead define these package dependencies within the infinispan feature group?

Comment (rhusar) [reply ee-feature-pack/galleon-common/src/main/resources/layers/standalone/web-clustering/layer-spec.xml]: @pferraro Turns out it is actually possible! Reworked.

**Review by pferraro — APPROVED** [review]

File: ee-feature-pack/galleon-common/src/main/resources/layers/standalone/infinispan/layer-spec.xml [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<layer-spec xmlns="urn:jboss:galleon:layer-spec:1.0" name="infinispan">
+
```
Comment (rhusar) [inline ee-feature-pack/galleon-common/src/main/resources/layers/standalone/infinispan/layer-spec.xml:25]: This should require `cdi`.

File: ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/infinispan/cdi/common/main/module.xml [hunk 15]
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ JBoss, Home of Professional Open Source.
+  ~ Copyright 2022, Red Hat Middleware LLC, and individual contributors
```
Comment (rhusar) [inline ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/infinispan/cdi/common/main/module.xml:4]: Stale copyright header.

File: ee-feature-pack/galleon-common/src/main/resources/layers/standalone/infinispan/layer-spec.xml (hunk 14 above)
Comment (rhusar) [reply ee-feature-pack/galleon-common/src/main/resources/layers/standalone/infinispan/layer-spec.xml:25]: Done.

File: ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/infinispan/cdi/common/main/module.xml (hunk 15 above)
Comment (rhusar) [reply ee-feature-pack/common/src/main/resources/modules/system/layers/base/org/infinispan/cdi/common/main/module.xml]: Fixed.

**Review by pferraro — APPROVED** [review]

**Comment by pferraro** [comment]
Ran a final sanity CI run and all looks good so far.  Will merge once complete.

**Pull request merged**